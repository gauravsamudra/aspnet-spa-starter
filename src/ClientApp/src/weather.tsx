import React, { useState, useEffect } from "react";

export interface IProps {
  id: number;
}

export interface IWeather {
  summary: string;
  temperatureC: number;
  date: string;
}

export function Weather(props: IProps) {
  const [weather, setWeather] = useState(null);

  async function fetchWeather(id) {
    const response = await fetch(`/WeatherForecast/${id}/`);
    setWeather(await response.json());
  }
  useEffect(() => {
    fetchWeather(props.id);
  }, [props.id]);

  if (!weather) {
    return <div>loading...</div>;
  }

  return (
    <div>
      <h2 className="summary">Forecast: {weather.summary}</h2>
      <strong>{weather.temperatureC}</strong> Â°C
      <br />
      Date: {weather.date}
    </div>
  );
}