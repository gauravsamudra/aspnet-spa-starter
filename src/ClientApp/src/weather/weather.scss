.Weather { 
    table {
        border-spacing: 0px;
        border-style: solid;
        border-width: 1px;
        border-color: #e0e0dc;
        
        td {
            padding-left: 18px;
            padding-right: 18px;
            padding-top: 6px;
            padding-bottom: 6px;
            border: 1px solid #e0e0dc;
        }
        th {
            padding-left: 18px;
            padding-right: 18px;
            padding-top: 6px;
            padding-bottom: 6px;
            border: 1px solid #e0e0dc;
            background-color: rgba(212,221,228,.5);
        }
        caption {
            padding-top: 6px;
            caption-side: bottom;
        }
    }    
}