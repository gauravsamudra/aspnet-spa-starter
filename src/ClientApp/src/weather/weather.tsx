import React, { useState, useEffect } from "react";
import './weather.scss'
export interface IProps {
}

export interface IWeather {
  summary: string;
  temperatureC: number;
  temperatureF: number;
  date: string;
}

export function Weather(props: IProps) {
  const [weatherForecasts, setWeatherForecasts] = useState([] as IWeather[]);

  async function fetchWeatherForecasts() {
    const response = await fetch('/WeatherForecast/');
    if (!response.ok) {
      return;
    }
    let weatherForecasts = await response.json();
    weatherForecasts.map(w => w.date = new Date(w.date).toDateString());
    setWeatherForecasts(weatherForecasts);
  }
  useEffect(() => {
    fetchWeatherForecasts();
  }, [props]);

  return (
    <div className="Weather">
      <table>
        <thead>
          <tr>
            <th>Date</th>
            <th>Temp. °C</th>
            <th>Temp. °F</th>
            <th>Summary</th>
          </tr>
        </thead>
        <tbody>
          {weatherForecasts.map((forecast, index) =>
            <tr key={index}>
              <td>{forecast.date}</td>
              <td>{forecast.temperatureC}</td>
              <td>{forecast.temperatureF}</td>
              <td>{forecast.summary}</td>
            </tr>
          )}
        </tbody>
        <caption>Weather forecast</caption>
      </table>
    </div>
  );
}