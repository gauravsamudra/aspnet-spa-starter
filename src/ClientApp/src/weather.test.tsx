import { render, unmountComponentAtNode } from "react-dom";
import React from "react";
import { act } from "react-dom/test-utils";

import { Weather, IWeather } from "./weather";

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

test('Weather compile with mock data', async() => {
  var id = 1;
  
  const expectedValue = {
    date: new Date('2019-04-05').toDateString(),
    temperatureC: -5,
    summary: 'Test London mild weather'
  } as IWeather;
  
  const fetchSpy = jest.spyOn(window, 'fetch').mockImplementation(() =>
    Promise.resolve({ 
      json: () => Promise.resolve(expectedValue)
    } as Response));
  await act(async() => {    
      render(<Weather id={id} />, container);
  });

  expect(fetchSpy.mock.calls.length).toBe(1);
  expect(container.innerHTML).toMatchSnapshot();

  fetchSpy.mockReset();
});