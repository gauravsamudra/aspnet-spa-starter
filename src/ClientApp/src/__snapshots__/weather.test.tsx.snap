// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather compile with mock data 1`] = `"<div><h2 class=\\"summary\\">Forecast: Test London mild weather</h2><strong>-5</strong> Â°C<br>Date: Fri Apr 05 2019</div>"`;
